### PR TITLE
Log shard movements

### PR DESCRIPTION
**Pull request by idegtiarenko**
This change adds a logger that logs such shard events as:
* starting
* relocating
* failing
* promoting from replica to primary

Such information might be useful when debugging shard allocation in cluster.
The logger needs to be enabled using `org.elasticsearch.cluster.routing.allocation.ShardChangesObserver:DEBUG`

Related to: #60747

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardChangesObserver.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.routing.RoutingChangesObserver;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.UnassignedInfo;
+
+public class ShardChangesObserver implements RoutingChangesObserver {
+
+    private static final Logger logger = LogManager.getLogger(ShardChangesObserver.class);
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardChangesObserver.java:19]: I would like to start a discussion here re the log level we should be logging the shard changes. I see the following possibilities:
* log all at info (currently implemented)
* log all at debug so you need to opt in to get those details (even though they might be very helpful in sum investigations)
* log all movements as log as we have few shards (may be less then 1000?)
* have include/exclude pattern (similarly to https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-network.html#transport-tracer)

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardChangesObserver.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Override
+    public void shardFailed(ShardRouting failedShard, UnassignedInfo unassignedInfo) {
+        logger.info("{} has failed on [{}]: {}", shardIdentifier(failedShard), failedShard.currentNodeId(), unassignedInfo.getReason());
+    }
+
+    @Override
+    public void replicaPromoted(ShardRouting replicaShard) {
+        logger.info("{} is promoted to primary on [{}]", shardIdentifier(replicaShard), replicaShard.currentNodeId());
+    }
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardChangesObserver.java:45]: This currently logs when shard:
* starts on the node
* starts relocating
* is promoted to primary
* fails

It does not log when shard starts initializing (for example after index creation or a snapshot restore)

Please let me know if you see any other scenarios that we should include.

I am thinking about logging something when we detect that shard can no longer stay on the current node nor move anywhere else in cluster, however it is a trickier change and I would like to do it in a separate pr.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java [hunk 3]
```diff
@@ -143,6 +139,20 @@ private RoutingAllocation(
         this.nodeReplacementTargets = nodeReplacementTargets(clusterState);
         this.desiredNodes = DesiredNodes.latestFromClusterState(clusterState);
         this.unaccountedSearchableSnapshotSizes = unaccountedSearchableSnapshotSizes(clusterState, clusterInfo);
+        this.routingChangesObserver = new RoutingChangesObserver.DelegatingRoutingChangesObserver(
+            isSimulating
```
Comment (arteam) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java:143]: Is there any particular reason we don't want log shard changes in the non-simulating mode?

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingChangesObserver.java [hunk 4]
```diff
@@ -25,7 +25,7 @@ default void shardStarted(ShardRouting initializingShard, ShardRouting startedSh
     /**
      * Called when relocation of a started shard is initiated.
      */
-    default void relocationStarted(ShardRouting startedShard, ShardRouting targetRelocatingShard) {}
+    default void relocationStarted(ShardRouting startedShard, ShardRouting targetRelocatingShard, String reason) {}
```
Comment (arteam) [inline server/src/main/java/org/elasticsearch/cluster/routing/RoutingChangesObserver.java:28]: Does it makes sense to create an enum to represent the reason for relocation? There should be a finite amount of reasons. We can make the method more type safe and less error-prone for typos.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java (hunk 3 above)
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java:143]: Yeap, simulating is a mode when we compute the desired balance allocation, not preform a real shard movements, so I do not think they are relevant for the user.

Here we would like to record actual shard and data movements that happened in the cluster.

File: server/src/main/java/org/elasticsearch/cluster/routing/RoutingChangesObserver.java (hunk 4 above)
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/routing/RoutingChangesObserver.java:28]: Yeah, I think this is a good idea

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardChangesObserver.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardChangesObserver.java:19]: I don't think we can reasonably log this volume at `INFO`. We may want to write it to a system index or data stream. For now I vote for `DEBUG`.

**Review by DaveCTurner — COMMENTED** [review]
Makes sense to me at `DEBUG`.

**Review by arteam — APPROVED** [review]
Still think we could use an enum to represent relocation reasons, but this looks good to me in the current form.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**